### PR TITLE
MIDI-94: Build training loop

### PR DESCRIPTION
**Pull request by JasiekKaczmarczyk**
[Build training loop](https://sacrebleu.atlassian.net/browse/MIDI-94)
[Create evaluation scripts](https://sacrebleu.atlassian.net/browse/MIDI-95)

Implemented training loop.

### CODE REVIEW

**Comment by JasiekKaczmarczyk** [comment]
Added dashboard to visualize embeddings in cls_token. You can provide list of strings separated by semicolon that will be queries for maestro dataset.

This model was trained for only one epoch so results might not be good.
Zaznaczenie_014

To run:
```python
streamlit run --server.port 4466 dashboards/cls_dashboard.py
```

It uses PCA to reduce embeddings to reasonable dimension before applying Decision Tree for classification.

**Comment by roszcz** [comment]
look cool! what was the MLM loss after one epoch? what's the f1 (or other classification metrics) on his classification?

**Comment by JasiekKaczmarczyk** [comment]
> look cool! what was the MLM loss after one epoch? what's the f1 (or other classification metrics) on his classification?

Here is one run. It was trained for 2 epochs with quantization (7, 3, 3). I will try to train this model some more but it takes really long.
https://wandb.ai/jasiekkaczmarczyk/masked-midi-modelling/runs/i0ep1w9s

**Comment by roszcz** [comment]
This is the core challenge of large machine learning projects - how to iterate quickly even though experiments are slow. If you're confident about the data flow, then I would start experimenting with hyperparameters - learning rate has probably the biggest impact, but you can also decrease sequence length or model size to make the training faster.

Please attach the run config to wandb, this will make it easier to investigate runs through dashboards:

```py
wandb.init(
    ...
    config=OmegaConf.to_container(cfg, resolve=True),
)
```

I also recommend using wandb sweeps to look for parameters that work better - as an example you can take a look at https://wandb.ai/roszcz/midi-bert/sweeps/xh64bbdk?workspace=user-roszcz - you can see that for some runs the models do not start to converge, while for others the loss starts to fall down after ingesting a couple million tokens:

image

File: dashboards/cls_dashboard.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
+import torch
+import numpy as np
+import seaborn as sns
+import streamlit as st
+import matplotlib.pyplot as plt
+from datasets import load_dataset
+from sklearn.decomposition import PCA
```
Comment (roszcz) [inline dashboards/cls_dashboard.py:9]: please keep the requirements.txt up to date :)

File: dashboards/cls_dashboard.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+    plot_confusion_matrix(cm)
+
+    return acc
+
+
+def main():
+    checkpoint = torch.load(
+        "checkpoints/masked-midi-modelling-2023-10-15-10-19-params-90.31M.ckpt"
+        # "checkpoints/pianomask_2023_10_08_19_10.pt"
+    )
```
Comment (roszcz) [inline dashboards/cls_dashboard.py]: Let's use some of the streamlit components to select one of the available checkpoints.

```py
from glob import glob

available_checkpoints = glob("checkpoints/*pt")
```

File: dashboards/cls_dashboard.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+    cfg = checkpoint["config"]
+    device = cfg.train.device
```
Comment (roszcz) [inline dashboards/cls_dashboard.py:125]: I think it would be better to map everything to CPU by default here, so it would be possible to run the dashboard while the model is training on the same machine 🤔

File: dashboards/cls_dashboard.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+            # shape [batch_size, seq_len, vocab_size]
+            out = model(input_ids=input_token_ids).logits
+            cls_embedding = out[:, 0, :]
+            outputs.append(cls_embedding)
+
+            for s in source:
+                for query in queries:
+                    if str.lower(query) in str.lower(s):
+                        labels.append(query)
```
Comment (roszcz) [inline dashboards/cls_dashboard.py]: It doesn't make sense to process everything and only later look for records matching the queries. Can you refactor it into steps more like `query -> dataset -> model -> PCA`?

File: dashboards/cls_dashboard.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+
+    assert len(X) == len(y)
+
+    plot_embeddings(X, labels)
+
+    # reducing dimensions
+    pca = PCA(n_components=128)
+    X_reduced = pca.fit_transform(X)
+
+    X_train, X_test, y_train, y_test = train_test_split(X_reduced, y, test_size=0.2, random_state=42)
```
Comment (roszcz) [inline dashboards/cls_dashboard.py:106]: If the data comes from a single maestro split, you are training more of a "piece recognition" pca, than a composer recognition. Can you try including the third maestro split in this classification test as well?

File: dashboards/cls_dashboard.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+    model_config = RobertaConfig(vocab_size=tokenizer.vocab_size)
+    model = RobertaForMaskedLM(model_config).to(device)
+
+    model.load_state_dict(checkpoint["model"])
+    # model.load_state_dict(checkpoint["model_state_dict"])
+    model.eval()
+
+    dataset_name = "JasiekKaczmarczyk/maestro-v1-sustain-masked"
+
+    queries_txt = st.text_input(label="Queries")
```
Comment (roszcz) [inline dashboards/cls_dashboard.py]:
```suggestion
    queries_txt = st.text_input(label="Queries", value="chopin, bach")
```

Please alway provide meaningful defaults for everything :)

File: dashboards/cls_dashboard.py [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+    queries = str.split(queries_txt, sep=";")
+
+    acc = evaluate_classification(
+        model, dataset_name=dataset_name, quantizer=quantizer, tokenizer=tokenizer, queries=queries, device=device
+    )
+    st.write(f"Classification accuracy is: {acc}")
```
Comment (roszcz) [inline dashboards/cls_dashboard.py:153]: Right now this generates:

image

But that's not enough to say if the classification is better than random guess - we don't know what's the dataset class balance.

**Review by roszcz — APPROVED** [review]
